cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandlerInternal.java

```diff
@@ -0,0 +1,94 @@
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import io.cdap.cdap.internal.app.store.RunRecordMeta;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * {@link HttpHandler} for serving internal program lifecycle REST API requests
+ */
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/namespaces/{namespace-id}")
+public class ProgramLifecycleHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
+
+  private final ProgramLifecycleService programLifecycleService;
+
+  @Inject
+  ProgramLifecycleHttpHandlerInternal(ProgramLifecycleService programLifecycleService) {
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  /**
+   * Return {@link RunRecordMeta} for the given program run id
+   *
+   * @param request {@link HttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespaceId namespace of the program
+   * @param appName name of the application
+   * @param appVersion version of the application
+   * @param type type of the program
+   * @param programName name of the program
+   * @param runid for which {@link RunRecordMeta} will be returned
+   * @throws Exception if failed to to get {@link RunRecordMeta}
+   */
+  @GET
+  @Path("/apps/{app-name}/versions/{app-version}/{program-type}/{program-name}/runs/{run-id}")
+  public void getProgramRunRecordMeta(HttpRequest request, HttpResponder responder,
+                                      @PathParam("namespace-id") String namespaceId,
+                                      @PathParam("app-name") String appName,
+                                      @PathParam("app-version") String appVersion,
+                                      @PathParam("program-type") String type,
+                                      @PathParam("program-name") String programName,
+                                      @PathParam("run-id") String runid) throws Exception {
+    ProgramType programType = getProgramType(type);
+    ProgramId programId = new ApplicationId(namespaceId, appName, appVersion).program(programType, programName);
+    RunRecordMeta runRecordMeta = programLifecycleService.getRun(programId.run(runid));
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(runRecordMeta));
+  }
+
+  /**
+   * Parses the give program type into {@link ProgramType} object.
+   *
+   * @param programType the program type to parse.
+   * @throws BadRequestException if the given program type is not a valid {@link ProgramType}.
+   */
+  private ProgramType getProgramType(String programType) throws BadRequestException {
```

Comment:
You don't need this function. You can just do:

```
ProgramType.valueOfCategoryName(programType, BadRequestException::new)
```

Reply:
Thanks. Done